ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/Jumping.hs

```diff
@@ -0,0 +1,294 @@
+      dynamoFragment <- fromJust <$> getDynamoFragment handlesVar
+      when (succWithOrigin (headSlot dynamoFragment) >= succWithOrigin lastJumpSlot + jumpSize) $ do
+        handles <- readTVar handlesVar
+        forM_ (map cschJumping $ Map.elems handles) $ \case
+          Jumper nextJumpVar _ Happy -> writeTVar nextJumpVar $ Just $! castPoint (headPoint dynamoFragment)
+          _ -> pure ()
+
+setJumpingState ::
+  ( MonadSTM m,
+    Ord peer
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  peer ->
+  ChainSyncJumpingState m blk ->
+  STM m ()
+setJumpingState handlesVar peer cschJumping =
+  modifyTVar handlesVar $ Map.update (\handle -> Just $! handle {cschJumping}) peer
+
+processJumpResult ::
+  ( MonadSTM m,
+    Ord peer, HasHeader blk, HasHeader (Header blk)
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  peer ->
+  JumpResult blk ->
+  STM m ()
+processJumpResult handlesVar peer jumpResult =
+  (cschJumping . (Map.! peer) <$> readTVar handlesVar) >>= \case
+    Dynamo _ -> pure ()
+    Objector _ -> pure ()
+    Jumper nextJumpVar goodPoint state ->
+        case (state, jumpResult) of
+          (FoundIntersection, _) ->
+            -- Ignore jump results when we already found the intersection.
+            pure ()
+          (Happy, AcceptedJump goodPoint') -> do
+            -- The jump was accepted; we bump the peer's candidate fragment to
+            -- the dynamo's candidate fragment up to the accepted point.
+            dynamoFragment <- fromJust <$> getDynamoFragment handlesVar
+            let slicedDynamoFragment = fst $ fromJust $ splitAfterPoint dynamoFragment goodPoint'
+            chainSyncStateVar <- cschState . (Map.! peer) <$> readTVar handlesVar
+            modifyTVar chainSyncStateVar $ \chainSyncState -> chainSyncState {csCandidate = slicedDynamoFragment}
+            setJumpingState handlesVar peer $ Jumper nextJumpVar goodPoint' Happy
+          (Happy, RejectedJump badPoint) ->
+            -- The previously happy peer just rejected a jump. We look for an
+            -- intersection.
+            lookForIntersection nextJumpVar goodPoint badPoint
+          (LookingForIntersection badPoint, AcceptedJump goodPoint') ->
+            -- The peer that was looking for an intersection accepts the given
+            -- point, helping us narrow things down.
+            lookForIntersection nextJumpVar goodPoint' badPoint
+          (LookingForIntersection _, RejectedJump badPoint') ->
+            -- The peer that was looking for an intersection rejects the given
+            -- point, helping us narrow things down.
+            lookForIntersection nextJumpVar goodPoint badPoint'
+  where
+    lookForIntersection nextJumpVar goodPoint badPoint = do
+      (dynamoFragment0 :: AnchoredFragment (Header blk)) <- fromJust <$> getDynamoFragment handlesVar -- full fragment
+      let dynamoFragment = fromJust $ sliceRangeExcl dynamoFragment0 goodPoint badPoint
```

Comment:
This `fromJust` call is going to fail if `goodPoint` is before the anchor of the dynamo fragment. e.g. if `goodPoint` is `GenesisPoint` and the fragment is not anchored at genesis.

Maybe this works better:
```
let dynamoAnchor = anchorPoint (anchor dynamoFragment0)
    dynamoFragment = fromJust $ do
      maxPoint <- maybeMax goodPoint dynamoAnchor
      sliceRangeExcl dynamoFragment0 maxPoint badPoint
```
where `dynamoAnchor` must be an ancestor or descendant of `goodPoint`, or otherwise it is fine to disconnect the jumper.

Reply:
Relatedly, the `fromJust` call might also fail if the dynamo makes a rollback.

If `badPoint` is not in the fragment, I think the last jump should be retried.